src/account/mod.rs

```diff
@@ -703,35 +745,42 @@ impl Account {
             .unwrap()
     }
 
-    /// Gets the account balance information.
-    pub fn balance(&self) -> AccountBalance {
-        let (incoming, outgoing) = self.list_messages(0, 0, Some(MessageType::Confirmed)).iter().fold(
-            (0, 0),
-            |(incoming, outgoing), message| {
-                if let Some(MessagePayload::Transaction(tx)) = message.payload() {
-                    let TransactionEssence::Regular(essence) = tx.essence();
-                    if !essence.internal() {
-                        if essence.incoming() {
-                            return (incoming + essence.value(), outgoing);
-                        } else {
-                            return (incoming, outgoing + essence.value());
+    pub(crate) async fn balance_internal(&self, sent_messages: &[Message]) -> AccountBalance {
+        let (incoming, outgoing) = self
+            .with_messages(|messages| {
+                messages.iter().filter(|m| m.confirmed == Some(true)).fold(
+                    (0, 0),
+                    |(incoming, outgoing), message: &MessageIndexation| {
+                        if message.internal == Some(false) {
+                            if message.incoming.unwrap_or(false) {
+                                return (incoming + message.value, outgoing);
+                            } else {
+                                return (incoming, outgoing + message.value);
+                            }
                         }
-                    }
-                }
-                (incoming, outgoing)
-            },
-        );
+                        (incoming, outgoing)
+                    },
+                )
+            })
+            .await;
+
         AccountBalance {
             total: self.addresses.iter().fold(0, |acc, address| acc + address.balance()),
             available: self
                 .addresses()
                 .iter()
-                .fold(0, |acc, addr| acc + addr.available_balance(&self)),
+                .fold(0, |acc, addr| acc + addr.available_balance(&sent_messages)),
             incoming,
             outgoing,
         }
     }
 
+    /// Gets the account balance information.
+    pub async fn balance(&self) -> crate::Result<AccountBalance> {
+        let sent_messages = self.list_messages(0, 0, Some(MessageType::Sent)).await?;
+        Ok(self.balance_internal(&sent_messages).await)
```

Comment:
How does this work when we don't have all messages because some gut pruned?

Reply:
it just gets what it can :/ it's something that Firefly agrees with.